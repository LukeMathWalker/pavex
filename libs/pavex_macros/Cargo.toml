[package]
name = "pavex_macros"
version = "0.1.0"
edition = "2021"

[lib]
proc-macro = true

[dependencies]
proc-macro2 = "1.0"
quote = "1.0"
syn = "2.0"

[dev-dependencies]
trybuild = "1.0"
pavex = { path = "../pavex" }
serde = { version = "1.0", features = ["derive"] }
