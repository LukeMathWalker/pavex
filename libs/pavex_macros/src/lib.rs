use proc_macro::TokenStream;

use quote::quote;
use syn::punctuated::Punctuated;
use syn::{parse_macro_input, Attribute, Data, DeriveInput, Error, GenericParam, Token};

#[allow(non_snake_case)]
#[proc_macro_attribute]
pub fn RouteParams(_metadata: TokenStream, input: TokenStream) -> TokenStream {
    let ast = parse_macro_input!(input as DeriveInput);

    if let Err(mut e) = reject_serde_attributes(&ast) {
        // We emit both the error AND the original struct.
        // This is useful to avoid spurious "the type doesn't exist anymore" additional errors
        e.extend(vec![TokenStream::from(quote! { #ast })]);
        return e;
    }

    let generics_with_bounds = &ast.generics;
    let generics_without_bounds = &ast
        .generics
        .params
        .iter()
        .map(|p| match p {
            GenericParam::Lifetime(l) => {
                let l = &l.lifetime;
                quote! { #l }
            }
            GenericParam::Type(t) => {
                let i = &t.ident;
                quote! { #i }
            }
            GenericParam::Const(c) => {
                let i = &c.ident;
                quote! { #i }
            }
        })
        .collect::<Punctuated<_, Token![,]>>();
    let mut attrs = ast.attrs.clone();
    attrs.push(syn::parse_quote!(#[derive(serde::Serialize, serde::Deserialize)]));
    let struct_name = &ast.ident;
    let expanded = quote! {
        #(#attrs)*
        #ast

        impl #generics_with_bounds pavex_runtime::serialization::StructuralDeserialize for #struct_name < #generics_without_bounds > {}
    };

    TokenStream::from(expanded)
}

fn reject_serde_attributes(ast: &DeriveInput) -> Result<(), TokenStream> {
    for attr in &ast.attrs {
        reject_serde_attribute(attr)?;
    }
    match &ast.data {
        Data::Struct(data) => {
            for field in &data.fields {
                for attr in &field.attrs {
                    reject_serde_attribute(attr)?;
                }
            }
        }
        Data::Enum(data) => {
            for variant in &data.variants {
                for attr in &variant.attrs {
                    reject_serde_attribute(attr)?;
                }
                for field in &variant.fields {
                    for attr in &field.attrs {
                        reject_serde_attribute(attr)?;
                    }
                }
            }
        }
        Data::Union(union) => {
            for field in &union.fields.named {
                for attr in &field.attrs {
                    reject_serde_attribute(attr)?;
                }
            }
        }
    }
    Ok(())
}

/// We don't want to allow `serde` attributes on the top-level struct or any of its fields,
/// because we rely on `serde`'s default behaviour to determine, at code-generartion time,
/// if the route params can be deserialized from the URL of the incoming request.
fn reject_serde_attribute(attr: &Attribute) -> Result<(), TokenStream> {
    let err_msg = "`RouteParams` does not support `serde` attributes on the top-level struct or any of its fields.\n\n\
      `RouteParams` takes care of deriving `serde::Serialize` and `serde::Deserialize` for your struct, using the default \
       configuration. This allow Pavex to determine, at code-generation time, if the route params can \
       be successfully extracted from the URL of incoming requests for the relevant routes (e.g. do you \
       have a named field that doesn't map to any of the registered route parameters?).\n\n\
       If the default `serde` configuration won't work for your case, you should not derive `RouteParams` and \
       opt instead for implementing `serde::Serialize` and `serde::Deserialize` directly for your struct (either \
       manually or using a derive with custom attributes).\nKeep in mind that by going down this route \
       you give up compile-time checking of the route parameters!";
    if let Some(ident) = attr.path().get_ident() {
        if ident == "serde" {
            return Err(Error::new_spanned(attr, err_msg)
                .into_compile_error()
                .into());
        }
    }
    Ok(())
}
