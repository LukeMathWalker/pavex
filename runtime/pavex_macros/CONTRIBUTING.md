# Testing

`pavex_macros` relies on [`trybuild`](https://github.com/dtolnay/trybuild) to verify that Pavex macros:

- Accept the expected input/parameter combinations
- Reject invalid input/parameter combinations

In the latter case, we also verify the returned error message to ensure that
diagnostics are accurate and actionable.

## Test filtering

`trybuild` is a custom test runner and it doesn't obey all `cargo test` built-in options.

For filtering, in particular, use

```bash
cargo t -- ui trybuild=not_a_function.rs
```

to run a subset of the whole UI test suite.
The `trybuild` argument doesn't have to be the filename: you can only specify a fragment of the path (e.g. `from_request`) to run
all tests in a specific sub-directory.

## Overwriting saved snapshots

To overwrite saved snapshots, set the `TRYBUILD=overwrite` environment variable:

```bash
# Overwrite all saved snapshots, if changed
TRYBUILD=overwrite cargo t 
# Overwrite snapshots for a subset of tests
TRYBUILD=overwrite cargo t -- ui trybuild=from_request
```
