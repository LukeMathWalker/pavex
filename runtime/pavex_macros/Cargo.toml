[package]
name = "pavex_macros"
version.workspace = true
description = "Procedural macros for the Pavex framwork."
keywords = ["pavex"]
edition.workspace = true
repository.workspace = true
license.workspace = true

[lib]
proc-macro = true

[features]
# Allow items annotated with our macros to be marked as `pub` without being
# reachable from outside the crate they were defined into.
# This is necessary to get doc-tests compiling in the `pavex` crate.
allow_unreachable_pub = []

[dependencies]
pavexc_attr_parser = { path = "../../compiler/pavexc_attr_parser", version = "=0.2.9" }
paste = { workspace = true }
darling = { workspace = true }
proc-macro2 = { workspace = true }
quote = { workspace = true }
syn = { workspace = true, features = ["visit-mut"] }
convert_case = { workspace = true }

[dev-dependencies]
trybuild = { workspace = true }
pavex = { path = "../pavex" }
serde = { workspace = true, features = ["derive"] }
