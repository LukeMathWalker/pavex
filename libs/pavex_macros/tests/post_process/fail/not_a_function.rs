use pavex_macros::post_process;

pub struct A;

impl A {
    #[post_process]
    const A: usize = 42;
}

fn main() {}
