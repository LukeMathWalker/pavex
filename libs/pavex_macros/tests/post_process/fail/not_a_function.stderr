error: #[pavex::post_process] can only be applied to functions and methods.
 --> tests/post_process/fail/not_a_function.rs:7:5
  |
7 |     const A: usize = 42;
  |     ^^^^^^^^^^^^^^^^^^^^
